### PR TITLE
69 model split

### PR DESCRIPTION
**Pull request by rogerkuou**
- [x] add `xr.Dataset` extension in MOT
- [x] function `is_valid` to check DS validity
- [x] function  `model_split` to split data for independent model trainings

### CODE REVIEW

**Comment by rogerkuou** [comment]
Hi @SarahAlidoost, when you have time, do you mind review this pull request?

In this PR I implemented the Xarray extension for MOT. There are two functions added: 
- `is_valid` to validate the Dataset
- `.mot.model_split` to perfor 1D or 2D model split

There is an[ example notebook](https://github.com/VegeWaterDynamics/motrainer/pull/78/files#diff-f5d59ebf3a41ac0fd070e907cf9d582549750379619d86aa92c1ab00dba395aa) to test this. The Zarr data needed is [here](https://nlesc-my.sharepoint.com/:f:/g/personal/o_ku_esciencecenter_nl/EqTUot-Wpe5Ci8EbuA1RHoAB3PXWmVm0eCUPPEKRoEHu4Q?e=t6Wqgk).

File: requirements.txt [hunk 1]
```diff
@@ -6,3 +6,5 @@ pandas
 scikit-optimize
 ipykernel
 matplotlib
+xarray
+dask
```
Comment (SarahAlidoost) [inline requirements.txt:10]: To use [xarray.open_zarr](https://docs.xarray.dev/en/stable/generated/xarray.open_zarr.html), [zarr](https://zarr.readthedocs.io/en/stable/installation.html#installation) should be installed too, right? :thinking:

File: motrainer/motrainer.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            bool: Result of check in Boolean.
+                  If all checks are passed, it will be True. Otherwise False.
+        """
+
+        flag_valid = True
+
+        # Dimension size should be 2
+        if len(self._obj.dims) != 2:
+            warnings.warn(
+                f'Dataset should have two dimensions: "space" and "time"', UserWarning
```
Comment (SarahAlidoost) [inline motrainer/motrainer.py:41]:
```suggestion
                'Dataset should have two dimensions: "space" and "time"', UserWarning
```

File: motrainer/motrainer.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            )
+            flag_valid = False
+
+        # space and time dimensions should exist
+        for dim in MOT_DIMS:
+            if not (dim in self._obj.dims):
```
Comment (SarahAlidoost) [inline motrainer/motrainer.py:47]:
```suggestion
            if dim not in self._obj.dims:
```

File: motrainer/motrainer.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            dask.bag: A Dask Databag of splited Datasets 
+        """
+
+        if isinstance(identifier, dict):
+            if len(identifier.keys())==0:
+                raise ValueError('identifier is empty')
+            if not set(identifier.keys()).issubset(MOT_DIMS):
+                raise ValueError('Acceptable keys are "space" and/or "time".')
+        elif isinstance(identifier, str):
+            if not (identifier in self._obj.variables):
```
Comment (SarahAlidoost) [inline motrainer/motrainer.py:77]:
```suggestion
            if identifier not in self._obj.variables:
```

File: motrainer/motrainer.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            identifier (dict | str): Split identifier. 
+            When `indentifier` is a dictionary, it should map "space" and/or "time" dimension with corresponding separation indentifier.
+            When `indentifier` is a string, `model_split` will use the corresponding field as the indertifier. This field can only be 1D.
+
+        Returns:
+            dask.bag: A Dask Databag of splited Datasets 
+        """
+
+        if isinstance(identifier, dict):
+            if len(identifier.keys())==0:
```
Comment (SarahAlidoost) [inline motrainer/motrainer.py:72]:
```suggestion
            if len(identifier.keys()) == 0:
```

File: motrainer/motrainer.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                raise ValueError('identifier is empty')
+            if not set(identifier.keys()).issubset(MOT_DIMS):
+                raise ValueError('Acceptable keys are "space" and/or "time".')
+        elif isinstance(identifier, str):
+            if not (identifier in self._obj.variables):
+                raise ValueError(f'Cannot find "{identifier}" in the Dataset')
+            if len(self._obj[identifier].dims)>1:
```
Comment (SarahAlidoost) [inline motrainer/motrainer.py:79]:
```suggestion
            if len(self._obj[identifier].dims) > 1:
```

File: motrainer/motrainer.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        # Check duplicated coordinates
+        for coord in self._obj.coords:
+            if np.unique(self._obj[coord]).shape != self._obj[coord].shape:
+                warnings.warn(f"Duplicated coordinates found in {coord}", UserWarning)
+                flag_valid = False
+
+        return flag_valid
+
+    def model_split(self, identifier: dict | str):
+        """Split a Dataset by indentifier for independent training tasks.
```
Comment (SarahAlidoost) [inline motrainer/motrainer.py:60]: I think it is confusing if the function name is `model_split` but the documentation says `"Split a Dataset"`. Can we rename the function to `dataset_split`? Also, can we say `Xarray Dataset` to make it more explicit?

**Review by SarahAlidoost — APPROVED** [review]
@rogerkuou nice work :+1: I could run the notebook and left some minor comments/suggestions here on changed files.

**Comment by rogerkuou** [comment]
Thanks @SarahAlidoost, I adapted to all your comments, will merge now.

**Pull request merged**